### ISSUE TITLE
Add user groups?

### ISSUE THREAD

**Issue by knoxsp** [hydraplatform/hydra-base#81]
Having multiple distinct user groups on the same system can become both a security issue and a usability problem. Should we support user groups natively in hydra, rather relying on an app to manage that stuff?

What behaviour should this exhibit? Have global users and non-global users

**Comment by pmslavin** [comment]
Are "groups" here different from the "roles" in the current `users`, `roles`, and `permissions` [model](https://github.com/hydraplatform/hydra-base/blob/master/hydra_base/lib/users.py)?

**Comment by knoxsp** [comment]
Yes. The idea with groups would be to sandbox projects and networks to a
defined set of users. For example, to avoid accidentally sharing a network
with a user from a different company.

**Comment by knoxsp** [comment]
Right, here's the idea (on the user-groups branch)

The crux of the matter is that some function calls like 'get_all_users' can return ALL system users. This is insecure and has the potentiial to cause issues of sharing (accidentally sharing with the wrong user).

Our decision is that only admins (or any user with the new **'view_users'** permission) can see all system users. Non-admins will see an emtpy list when requesting all users, while not in the scope of a group.

###Organisations
An **Organisation** is the highest level. This must be created by an admin.
An organisation can have **members**
Members of an organisation can have **roles**. These are the same roles as global, except they only apply in the context of an organisation. There are additional permissions, such as 'create group' which are not applicable globally. Conversely, 'create organisation' doesn't make sense within an organisation.

Organisation members with the appropriate role/permission can add and remove members, but they can still not list all users, so this invitation must happen by inviting a member by explicitly stating their username / email addresss, and that member accepting the invitation. ex
```
invite_member_go_organisation(organisation_id, username='knoxsp')
```
and
```
accept_organisation_invitation(organisation_id)
```

### Group Management
Once in an organisation, a user must be added to a group (either through creating it themselves or being added to it). 

A user (again, with the appropriate permissions) can create a project within this group. When they wish to share the project, they can call `get_all_users(group_id)`. This will first check that the user is in the group, and with the correct permissions. It will then return only the users in this group.

Adding & removing a user to a group happens at an organisation level --- an organisation with the correct permissions will call `add_user_to_group(group_id, user_id)` and `remove_user_from_group(group_id, user_id)` respectively

**Comment by pmslavin** [comment]
That seems a good compromise between allowing easy exploration within a confined namespace, but also allowing interaction with unrelated users from the global namespace.

We mentioned (offline, so not above) that users may also have personal projects, which they can presumably share with others - so are these personal projects effectively user-specific groups for which the user in question is an administrator (and implemented as such), or is the set of a user's personal projects some other entity (and therefore implemented separately from groups)?

**Comment by knoxsp** [comment]
I hope to leave the existing management of projects alone, so there would be no explicit implementation for a single user. A user can create a project without being in an organisation. But they can't search for users with whom to share the project. These would need to be invited explicitly. (they would need to know the username)

One further thought: Perhaps this restriction could be configurable. Perhaps some deployments of hydra will not want such restrictions in place.

**Comment by pmslavin** [comment]
Having this configurable sounds attractive; some users may just be running a local instance for their own use, while the public servers need to support multiple organisations, etc.

**Comment by knoxsp** [comment]
After some discussion a new, more general, approach to this is being considered.
In this, there is only one concept of user group, which can contain multiple other user groups.
A group 'type' identifies what kind of group it is. So rather than having an 'organisation', we have a group of `type` organisation. 

As with the previous idea, users can have roles assigned within groups.
